subprojects/base-services/src/main/java/org/gradle/internal/reflect/DirectInstantiator.java

```diff
@@ -25,6 +25,23 @@
 import java.util.Arrays;
 
 public class DirectInstantiator implements Instantiator {
+    /*
+     * Every time we look up a constructor, we need to pass its types as a class array. Reusing the most common
+     * argument list sizes avoids a lot of garbage.
+     */
+    private static final ThreadLocal<Class[][]> REUSABLE_CLASS_ARRAYS = new ThreadLocal<Class[][]>() {
```

Comment:
Did you actually measure an improvement with this change? I'm really reluctant on adding thread locals, as they may create memory leaks (here, we would keep strong references onto `Class` instances). Also we're creating a link between a long lived object (the thread local) and short lived objects (the class instances), which should slow down GC.

:x: